### PR TITLE
Add view-only boot order component to VM details dialog

### PR DESCRIPTION
**Pull request by pcbailey**
This PR is to add a boot order component to the VM details dialog. Editing functionality will be added in a separate PR in the future.

Depends on: https://github.com/kubevirt/web-ui/pull/207

### CODE REVIEW

File: src/k8s/request.js [hunk 1]
```diff
@@ -362,6 +362,33 @@ const getDevices = vm => {
...
+  const disks = getDisks(vm);
+  const nics = getInterfaces(vm);
+  let largestIdx = -1;
+
+  disks.forEach(disk => {
+    const bootOrder = get(disk, 'bootOrder', -1);
+    largestIdx = bootOrder > largestIdx ? bootOrder : largestIdx;
+  });
+
+  nics.forEach(nic => {
```
Comment (suomiy) [inline src/k8s/request.js]: I would rather not set explicitly boot order for networks when creating VMs. Except the ones the user chooses. I know this concerns editing the VMs detail, so maybe we should have an option to discard some nics (or disks) from the boot order? Otherwise this behavior could have security ramifications.

Comment (suomiy) [reply src/k8s/request.js]: maybe there could be one component for boot order which could be used in VM details and in the create VM dialog? It could have its own step (something like BIOS?). So we would not be forced to set bootorder to all devices - just the ones user chooses. Thoughts?

Comment (pcbailey) [reply src/k8s/request.js]: > maybe there could be one component for boot order which could be used in VM details and in the create VM dialog? It could have its own step (something like BIOS?). So we would not be forced to set bootorder to all devices - just the ones user chooses. Thoughts?

+1... I like this idea. @jelkosz @rawagner What do you think?

Comment (rawagner) [reply src/k8s/request.js]: sounds good to me!

File: sass/components/VmDetails/boot-order.scss [hunk 2]
```diff
@@ -0,0 +1,4 @@
+.kubevirt-boot-order__ol {
```
Comment (rawagner) [inline sass/components/VmDetails/boot-order.scss]: I think that its incorrect to use html tag `ol` in css selector. Correct would be something like `.kubevirt-boot-order__list`

File: src/components/Details/BootOrder/BootOrder.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+export const BootOrder = props => {
+  const { bootableDevices } = props;
+  const sortedBootableDevices = bootableDevices.sort((a, b) => a.bootOrder - b.bootOrder);
+  const listItems = [];
```
Comment (rawagner) [inline src/components/Details/BootOrder/BootOrder.js]: can be reduced to `listItems = sortedBootableDevices.map((dev, idx) => <li key={`${idx}-${dev.name}`}>{dev.name}</li>)`;

File: src/k8s/request.js (hunk 1 above)
Comment (suomiy) [reply src/k8s/request.js]: After a discussion, it was decided that everything would be bootable. The boot order will be the order of devices in CreateDialog. Afterwards, it can be edited in VM detail with possibility to change or remove the bootOrder.

Comment (suomiy) [reply src/k8s/request.js]: Boot order should be set when creating VMs. Created relevant issue: #230

File: src/components/Details/BootOrder/BootOrder.js (hunk 3 above)
Comment (pcbailey) [reply src/components/Details/BootOrder/BootOrder.js]: Nice and clean! Done!

File: sass/components/VmDetails/boot-order.scss (hunk 2 above)
Comment (pcbailey) [reply sass/components/VmDetails/boot-order.scss]: Ah... I wasn't aware of that. Thanks! Done!

File: src/k8s/vmBuilder.js [hunk 4]
```diff
@@ -282,3 +282,86 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
...
+
+  // assigned indexes start at two as the first index is assigned directly by the user
+  return largestIdx !== -1 ? largestIdx + 1 : BOOT_ORDER_SECOND;
+};
+
+export const sequentializeBootOrderIndexes = vm => {
+  const sortedBootableDevices = getBootableDevicesInOrder(vm);
+
+  let index;
+  for (index = 0; index <= sortedBootableDevices.length; index++) {
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: .forEach((device, index) => {}) is more readable IMO

File: src/k8s/vmBuilder.js [hunk 5]
```diff
@@ -282,3 +282,86 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
...
+  });
+
+  nics.forEach(nic => {
+    const bootOrder = get(nic, 'bootOrder', null);
+    if (bootOrder) {
+      bootableDevices.push(nic);
+    }
+  });
+
+  return bootableDevices;
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: please use shorthands 

`return [..nics.filter(predicate), ...disks.filter(predicate)]`

File: src/k8s/vmBuilder.js [hunk 6]
```diff
@@ -282,3 +282,86 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
...
+};
+
+export const getBootableDevicesInOrder = vm => getBootableDevices(vm).sort((a, b) => a.bootOrder - b.bootOrder);
+
+export const isNic = device => !!get(device, 'bridge', null);
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: I do not think this is good enough to identify disks/nics. Everything except name is optional (https://kubevirt.io/api-reference/master/definitions.html#_v1_interface). Please restructure the code so this is not needed or make a wrapper object.

File: src/k8s/vmBuilder.js [hunk 7]
```diff
@@ -282,3 +282,86 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
...
+
+export const isDisk = device => !!get(device, 'disk', null);
+
+export const updateBootOrder = (vm, device, bootOrder) => {
+  if (isNic(device)) {
+    getInterfaces(vm).forEach(nic => {
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: why do we have to find that device again anyway? isn't nic and device the same exact object?

File: src/k8s/request.js [hunk 8]
```diff
@@ -193,6 +194,8 @@ export const createVm = async (k8sCreate, templates, basicSettings, networks, st
 
   const vm = selectVm(processedTemplate.objects);
 
+  sequentializeBootOrderIndexes(vm);
```
Comment (suomiy) [inline src/k8s/request.js]: I think this should be also applied for createVmTemplate

File: src/k8s/vmBuilder.js [hunk 9]
```diff
@@ -282,3 +282,86 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
...
+  const disks = getDisks(vm);
+  const nics = getInterfaces(vm);
+  let largestIdx = -1;
+
+  disks.forEach(disk => {
+    const bootOrder = get(disk, 'bootOrder', -1);
+    largestIdx = bootOrder > largestIdx ? bootOrder : largestIdx;
+  });
+
+  nics.forEach(nic => {
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: could be also written as `Math.max(...nics.map())`

File: src/k8s/vmBuilder.js [hunk 10]
```diff
@@ -18,9 +18,9 @@ export const addDisk = (vm, defaultDisk, storage, getSetting) => {
   };
   if (storage.isBootable) {
     const imageSource = getSetting(PROVISION_SOURCE_TYPE_KEY);
-    diskSpec.bootOrder = imageSource === PROVISION_SOURCE_PXE ? BOOT_ORDER_SECOND : BOOT_ORDER_FIRST;
+    diskSpec.bootOrder = imageSource === PROVISION_SOURCE_PXE ? assignBootOrderIndex(vm) : BOOT_ORDER_FIRST;
   } else {
-    delete diskSpec.bootOrder;
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: I think this should not include cloud init disks.

File: src/utils/patches.js [hunk 11]
```diff
@@ -387,3 +392,46 @@ export const addPrefixToPatch = (prefix, patch) => ({
   ...patch,
   path: `${prefix}${patch.path}`,
 });
+
+const getDeviceType = device => {
+  let type;
+  if (isNic(device)) {
```
Comment (suomiy) [inline src/utils/patches.js]: again, we should already know here what kind of device we have

Comment (suomiy) [reply src/utils/patches.js]: I guess easiest way is to wrap the disks/nic in a object with type attribute.

File: src/utils/tests/patches.test.js [hunk 12]
```diff
@@ -378,34 +380,39 @@ describe('utils.js tests', () => {
...
     const vmWithoutNetworks = getVM(false);
 
-    let patch = getAddNicPatch(vmWithoutNetworks, nic);
+    const patch = getAddNicPatch(vmWithoutNetworks, nic);
     expect(patch).toHaveLength(2);
     comparePatch(patch[0], '/spec/template/spec/domain/devices/interfaces/0', intface);
     comparePatch(patch[1], '/spec/template/spec/networks', [network]);
-
+  });
+  it('AddNicPatch - VM without interfaces', () => {
```
Comment (suomiy) [inline src/utils/tests/patches.test.js]: +1 for splitting the tests

File: src/components/Details/VmDetails/VmDetails.js [hunk 13]
```diff
@@ -231,6 +234,9 @@ export class VmDetails extends React.Component {
 
                   <dt>Pod</dt>
                   <dd>{PodResourceLink ? <PodResourceLink /> : DASHES}</dd>
+
+                  <dt>Boot Order</dt>
```
Comment (suomiy) [inline src/components/Details/VmDetails/VmDetails.js]: this could make a sense also for VmTemplateDetails

File: src/k8s/vmBuilder.js (hunk 4 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: Done!

File: src/k8s/vmBuilder.js (hunk 5 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: Done!

File: src/k8s/vmBuilder.js (hunk 6 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: Done!

File: src/k8s/vmBuilder.js (hunk 7 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: This was me being stupid. Thanks for making me realize it. It allowed me to clean things up significantly.

File: src/utils/patches.js (hunk 11 above)
Comment (pcbailey) [reply src/utils/patches.js]: Done!

File: src/utils/tests/patches.test.js (hunk 12 above)
Comment (pcbailey) [reply src/utils/tests/patches.test.js]: =)

File: src/components/Details/VmDetails/VmDetails.js (hunk 13 above)
Comment (pcbailey) [reply src/components/Details/VmDetails/VmDetails.js]: Agreed! Done!

File: src/constants/index.js [hunk 14]
```diff
@@ -17,6 +17,9 @@ export const KUBEVIRT_API_VERSION = 'kubevirt.io/v1alpha3';
 export const OS_LABEL = 'kubevirt.io/os';
 export const POD_NETWORK = 'Pod Networking';
 
+export const DEVICE_TYPE_INTERFACE = 'interfaces';
```
Comment (suomiy) [inline src/constants/index.js]: I know this is just an id but does it have to be plural?

File: src/k8s/vmBuilder.js [hunk 15]
```diff
@@ -282,3 +289,43 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
...
+  }
+  return bootOrder;
+};
+
+const getNextAvailableBootOrderIndex = vm => {
+  const sortedBootableDevices = getBootableDevicesInOrder(vm);
+  const numBootableDevices = sortedBootableDevices.length;
+
+  // assigned indexes start at two as the first index is assigned directly by the user
+  return numBootableDevices > 0 ? sortedBootableDevices[numBootableDevices - 1].value.bootOrder + 1 : BOOT_ORDER_SECOND;
```
Comment (suomiy) [inline src/k8s/vmBuilder.js]: can we return 0 instead of BOOT_ORDER_SECOND ? and do this when we use it : 

`getNextAvailableBootOrderIndex(vm) || BOOT_ORDER_SECOND`

this way, the function will contain less magic

File: src/utils/patches.js [hunk 16]
```diff
@@ -387,3 +392,50 @@ export const addPrefixToPatch = (prefix, patch) => ({
   ...patch,
   path: `${prefix}${patch.path}`,
 });
+
+const getWrappedDeviceType = device => {
+  let type;
+  if (isNic(device)) {
+    type = DEVICE_TYPE_INTERFACE;
+  } else if (isDisk(device)) {
```
Comment (suomiy) [inline src/utils/patches.js]: I know this was needed historically but `isNic` and `isDisk` are not needed anymore. You can just call `get(device, 'type')` instead of this method,

File: src/utils/patches.js [hunk 17]
```diff
@@ -387,3 +392,50 @@ export const addPrefixToPatch = (prefix, patch) => ({
...
+      const deviceType = getWrappedDeviceType(device);
+      const currDevIndex = getDeviceIndex(vm, device.value.name, deviceType);
+      // Account for index change once device is removed
+      const deviceIndex =
+        deviceType === removedDeviceType && currDevIndex > removedDeviceIndex ? currDevIndex - 1 : currDevIndex;
+
+      if (deviceIndex !== -1 && deviceType) {
+        const patch = {
+          op: 'replace',
+          path: `/spec/template/spec/domain/devices/${deviceType}/${deviceIndex}/bootOrder`,
```
Comment (suomiy) [inline src/utils/patches.js]: Oh! I see now, why is that needed. I think it would be probably better to decouple the type from the path fragment. I would rather use the singular for type and then do the mapping of the type to the path fragment. Quick and readable way is to use object for mappings for example.

this way, the functionality will not break when somebody decides to refactor the type

File: src/constants/index.js (hunk 14 above)
Comment (suomiy) [reply src/constants/index.js]: please look at the comment in `patches.js`

File: src/utils/tests/patches.test.js [hunk 18]
```diff
@@ -151,6 +152,7 @@ const podNic = {
 
 const intface = {
```
Comment (suomiy) [inline src/utils/tests/patches.test.js]: btw, does it have to be called intface ?

File: src/k8s/vmBuilder.js (hunk 10 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: Done!

File: src/constants/index.js (hunk 14 above)
Comment (pcbailey) [reply src/constants/index.js]: Agreed and done!

File: src/utils/patches.js (hunk 16 above)
Comment (pcbailey) [reply src/utils/patches.js]: Good point. Done!

File: src/utils/patches.js (hunk 17 above)
Comment (pcbailey) [reply src/utils/patches.js]: Agreed and done! Could you please check the way I implemented your suggestion to make sure this is what you were thinking?

File: src/utils/tests/patches.test.js (hunk 18 above)
Comment (pcbailey) [reply src/utils/tests/patches.test.js]: I don't believe I named this variable, but I'm pretty sure the reason they went with "intface" instead of "interface" is because "interface" is a reserved keyword in JavaScript. I typically use "iface" when shortening it, but it's not much better. Any suggestions?

File: src/k8s/vmBuilder.js (hunk 15 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: I think it makes more sense as it's currently written. The function is just expected to return the next available index, so the caller wouldn't expect to have to handle a special case like that when using it. I just reused the existing constant to avoid having a magic number and to increase readability. 

I tweaked the last line a small bit to make it a bit more understandable at first glance. Thoughts?

File: src/utils/patches.js (hunk 17 above)
Comment (suomiy) [reply src/utils/patches.js]: yup, looks good! 

Although I would rather name it DEVICE_TYPE_INTERFACE or something like that. INTERFACE/DISK seems a bit generic.

File: src/k8s/vmBuilder.js (hunk 15 above)
Comment (suomiy) [reply src/k8s/vmBuilder.js]: ok, your call :) 

We have just one caller of the method and I would rather move the hacky logic higher up the chain. I was more focused on the method's name (`getNextAvailableBootOrderIndex`) which doesn't adhere to the reality. 
Nevertheless both are valid approaches.

File: src/utils/tests/patches.test.js (hunk 18 above)
Comment (suomiy) [reply src/utils/tests/patches.test.js]: I see, I forgot about that. You can leave it then, if you want..

Comment (pcbailey) [reply src/utils/tests/patches.test.js]: I'll update it in the future if I think of a better alternative.

File: src/utils/patches.js (hunk 17 above)
Comment (pcbailey) [reply src/utils/patches.js]: Agreed and done! =)

File: src/k8s/vmBuilder.js (hunk 15 above)
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: Understood. I think the name is valid, though. All the hacky logic does is keep the method from returning an index of 1, which isn't technically an available index since it's assigned explicitly. I think I'll leave it as-is for the time being. We can always change it later if it becomes problematic.

Comment (suomiy) [reply src/k8s/vmBuilder.js]: ok

File: src/k8s/tests/request.test.js [hunk 19]
```diff
@@ -87,9 +87,9 @@ const testDataVolumeStorage = (
 
 const testStorage = (vm, storageIndex, bootOrder, expectedName) => {
   expect(vm.spec.template.spec.domain.devices.disks[storageIndex].name).toBe(expectedName);
-  expect(vm.spec.template.spec.domain.devices.disks[storageIndex].bootOrder).toBe(
-    bootOrder !== -1 ? bootOrder : undefined
-  );
+  vm.spec.template.spec.domain.devices.disks.forEach(disk => {
```
[the remark this answers is not in the document]
Comment (pcbailey) [reply src/k8s/tests/request.test.js]: Done!

File: src/components/Details/VmDetails/fixtures/VmDetails.fixture.js [hunk 20]
```diff
@@ -23,7 +23,27 @@ export const vmFixtures = {
...
+        spec: {
+          domain: {
+            devices: {
+              rng: {},
+              interfaces: [
+                { bridge: {}, name: 'podNetworkName', bootOrder: 2 },
+                { bridge: {}, name: 'pxeNetworkName', bootOrder: 1 },
+              ],
+              disks: [
+                { disk: {}, name: 'disk-one', bootOrder: 3 },
```
Comment (suomiy) [inline src/components/Details/VmDetails/fixtures/VmDetails.fixture.js]: can you add a cloud init to the vm ? so that we can see the snapshot does not contain it

File: src/utils/tests/patches.test.js [hunk 21]
```diff
@@ -436,3 +456,45 @@ describe('utils.js tests', () => {
...
+  const interfaces = [
+    { bridge: {}, name: 'nic-one', bootOrder: 4 },
+    { bridge: {}, name: 'nic-two', bootOrder: 5 },
+    { bridge: {}, name: 'nic-three', bootOrder: 6 },
+  ];
+
+  let vm;
+
+  beforeEach(() => {
+    vm = cloneDeep(getVM(false));
```
Comment (suomiy) [inline src/utils/tests/patches.test.js]: doesn't this function already return brand new object?

File: src/utils/tests/patches.test.js [hunk 22]
```diff
@@ -436,3 +456,45 @@ describe('utils.js tests', () => {
     comparePatch(patch[0], '/spec', { running: false });
   });
 });
+
+describe('getDeviceBootOrderPatch() tests', () => {
+  const disks = [
+    { disk: {}, name: 'disk-one', bootOrder: 1 },
+    { disk: {}, name: 'disk-two', bootOrder: 2 },
+    { disk: {}, name: 'disk-three', bootOrder: 3 },
```
Comment (suomiy) [inline src/utils/tests/patches.test.js]: can you please add the non bootable cloud init here as well?

File: src/k8s/vmBuilder.js [hunk 23]
```diff
@@ -18,14 +26,21 @@ export const addDisk = (vm, defaultDisk, storage, getSetting) => {
   };
   if (storage.isBootable) {
     const imageSource = getSetting(PROVISION_SOURCE_TYPE_KEY);
-    diskSpec.bootOrder = imageSource === PROVISION_SOURCE_PXE ? BOOT_ORDER_SECOND : BOOT_ORDER_FIRST;
+    diskSpec.bootOrder = imageSource === PROVISION_SOURCE_PXE ? assignBootOrderIndex(vm) : BOOT_ORDER_FIRST;
+  } else if (!isCloudInitDisk(vm, diskSpec)) {
```
[the remark this answers is not in the document]
Comment (pcbailey) [reply src/k8s/vmBuilder.js]: You're definitely right. I should have caught that. Thanks! =)

File: src/components/Details/VmDetails/fixtures/VmDetails.fixture.js (hunk 20 above)
Comment (pcbailey) [reply src/components/Details/VmDetails/fixtures/VmDetails.fixture.js]: Done!

File: src/utils/tests/patches.test.js (hunk 21 above)
Comment (pcbailey) [reply src/utils/tests/patches.test.js]: It does... lol. Fixed!

File: src/utils/tests/patches.test.js (hunk 22 above)
Comment (pcbailey) [reply src/utils/tests/patches.test.js]: Done!

File: src/components/Details/BootOrder/BootOrder.js [hunk 24]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+export const BootOrder = props => {
+  const { bootableDevices } = props;
```
Comment (mareklibra) [inline src/components/Details/BootOrder/BootOrder.js]: Nitpicking: Rest of the code follows syntax:
```
const MyComponent = ({ bootableDevices }) => {};
```

Comment (pcbailey) [reply src/components/Details/BootOrder/BootOrder.js]: Fixed and rebased. =)
